src/control/server/ctl_system.go

```diff
@@ -83,15 +47,74 @@ func (svc *ControlService) SystemMemberQuery(ctx context.Context, req *ctlpb.Sys
 		return nil, err
 	}
 
-	svc.log.Debug("received SystemMemberQuery RPC; reporting DAOS system members")
+	svc.log.Debugf("received SystemMemberQuery RPC; reporting DAOS system members (%+v)", svc.membership.GetMembers())
 
-	resp.Members = common.MembersToPB(svc.membership.GetMembers())
+	membersPB, err := common.MembersToPB(svc.membership.GetMembers())
+	if err != nil {
+		return nil, err
+	}
+	resp.Members = membersPB
 
 	svc.log.Debug("responding to SystemMemberQuery RPC")
 
 	return resp, nil
 }
 
+func (svc *ControlService) memberStop(ctx context.Context, leader *IOServerInstance,
+	member *common.SystemMember) *common.SystemMemberResult {
+
+	result := common.NewSystemMemberResult(member.String(), "stop")
+
+	svc.log.Debugf("MgmtSvc.systemStop kill member %+v\n", *member)
+
+	// TODO: force should be applied if a number of retries fail
+	resp, err := leader.msClient.Stop(ctx, member.Addr.String(),
+		&mgmtpb.KillRankReq{Rank: member.Rank, Force: false})
+	if err != nil {
+		result.Err = err
+	} else if resp.GetStatus() != 0 {
+		result.Err = errors.Errorf("DAOS returned error code: %d\n",
+			resp.GetStatus())
+	}
+
+	if result.Err == nil {
+		svc.membership.Remove(member.Rank)
+	} else {
+		svc.log.Debugf("MgmtSvc.systemStop error %s\n", result.Err)
+	}
+
+	return result
+}
+
+// systemStop sends KillRank gRPC requests to system membership list.
+func (svc *ControlService) systemStop(ctx context.Context, leader *IOServerInstance) (common.SystemMemberResults, error) {
+	members := svc.membership.GetMembers()
+	results := make(common.SystemMemberResults, 0, len(members))
+
+	leaderMember, err := members.GetMember(uint32(*leader.getSuperblock().Rank))
+	if err != nil {
+		return nil, err
+	}
+
+	// total retry timeout, allows for 10 retries
+	ctx, _ = context.WithTimeout(ctx, stopTimeout)
+
+	// TODO: inhibit rebuild on pool services, parallelise and make async.
+	for _, rank := range members.GetRanks() {
+		member := members[rank]
```

Comment:
This is an unsafe map access. Wouldn't it be better to iterate over the slice returned by members.GetMembers() ?